xwiki-platform-core/xwiki-platform-web/xwiki-platform-web-templates/src/test/java/org/xwiki/web/MacrosTest.java

```diff
@@ -287,4 +297,78 @@ void displayUser() throws Exception
             + "</div>", out.toString().trim());
     }
+
+    @Test
+    void getSanitizedURLAttributeValue() throws Exception
+    {
+        HTMLScriptService htmlScriptService = mock(HTMLScriptService.class);
+        this.oldcore.getMocker().registerComponent(ScriptService.class, "html", htmlScriptService);
+        SecurityScriptService securityScriptService = mock(SecurityScriptService.class);
+        this.oldcore.getMocker().registerComponent(ScriptService.class, "security", securityScriptService);
+        URLSecurityScriptService urlSecurityScriptService = mock(URLSecurityScriptService.class);
+        this.oldcore.getMocker().registerComponent(ScriptService.class, "security.url", urlSecurityScriptService);
+        when(securityScriptService.get("url")).thenReturn(urlSecurityScriptService);
+
+        String htmlElement = "input";
+        String attributeName = "value";
+        String url = "javascript:alert('foobar')";
+        String fallbackUrl = "xwiki/bin/view/Sandbox";
+        String script = "#getSanitizedURLAttributeValue($myHtmlElement,$myAttributeName,$myUrl,$myFallback,"
+            + "$sanitizedResult)";
+
+        VelocityContext velocityContext = this.velocityManager.getVelocityContext();
+        velocityContext.put("myHtmlElement", htmlElement);
+        velocityContext.put("myAttributeName", attributeName);
+        velocityContext.put("myUrl", url);
+        velocityContext.put("myFallback", fallbackUrl);
+
+        URI fallBackUri = new URI(fallbackUrl);
+        when(urlSecurityScriptService.parseToSafeURI(url)).thenReturn(null);
+        when(urlSecurityScriptService.parseToSafeURI(fallbackUrl)).thenReturn(fallBackUri);
+        when(htmlScriptService.isAttributeSafe(htmlElement, attributeName, fallBackUri.toString())).thenReturn(true);
+
+        // First check: given URL is not secure according to security.url, so we fallback and the HTMLScriptService
+        // allows the fallback URL.
+        StringWriter out = new StringWriter();
+        this.velocityManager.evaluate(out, "getSanitizedURLAttributeValue", new StringReader(script));
+        verify(urlSecurityScriptService).parseToSafeURI(url);
+
+        assertEquals(fallbackUrl, velocityContext.get("sanitizedResult"));
+
+        // Second check: given URL is not secure according to security.url, so we fallback but the HTMLScriptService
+        // does not allow the fallback URL.
+        when(htmlScriptService.isAttributeSafe(htmlElement, attributeName, fallBackUri.toString())).thenReturn(false);
+        out = new StringWriter();
+        this.velocityManager.evaluate(out, "getSanitizedURLAttributeValue", new StringReader(script));
+        verify(htmlScriptService, times(2)).isAttributeSafe(htmlElement, attributeName, fallBackUri.toString());
+        assertEquals("", velocityContext.get("sanitizedResult"));
+
+        // Third check: both the given URL and the fallback URL are not secure according to security.url
```

Comment:
Unless the checks are dependent of each other, this feel like each check should be in its own test method.

Reply:
Old debate of structure of tests :) Here my argument to have a single method is that old my checks share the same fixture, and also the current structure of the test class is basically one method = one macro being tested. So I decided to follow the same.

Reply:
The only pro I see to a methods splitting is that you get more failures at once (and finer grained), but I won't fight for this :)

Reply:
You can get the same effect by using `assertAll`, see https://junit.org/junit5/docs/current/user-guide/#writing-tests-assertions.